src/sources/soundsourceproxy.cpp

```diff
@@ -422,97 +426,119 @@ void SoundSourceProxy::updateTrackFromSource(
+        // Continue with nullptr
+        pCoverImg = nullptr;
     }
 
     // Parse the tags stored in the audio file
-    const auto metadataImported =
+    auto metadataImported =
             m_pSoundSource->importTrackMetadataAndCoverImage(
                     &trackMetadata, pCoverImg);
-    if (metadataImported.first != mixxx::MetadataSource::ImportResult::Succeeded) {
-        kLogger.warning() << "Failed to parse track metadata and/or cover art from file"
-                   << getUrl().toString();
-        return; // abort
-    }
-
-    if (!metadataSynchronized &&
-            (trackMetadata.getTrackInfo().getArtist().isEmpty() ||
-                    trackMetadata.getTrackInfo().getTitle().isEmpty())) {
-        // Fallback: If Artist or title fields are blank initially try to parse
-        // them from the file name.
-        // TODO(rryan): Should we re-visit this decision?
-        parseMetadataFromFileName(&trackMetadata, m_pTrack->getFileInfo().fileName());
+    if (metadataImported.first == mixxx::MetadataSource::ImportResult::Failed) {
+        kLogger.warning()
+                << "Failed to import track metadata"
+                << (pCoverImg ? "and embedded cover art" : "")
+                << "from file"
+                << getUrl().toString();
+        // Continue for now, but the abort may follow shortly if the
+        // track already has metadata (see below)
     }
-
-    // Until this point the track object has not been modified.
-    // Now we start updating it...
-
-    // Initialize or update the file/format as reported by the
-    // responsible SoundSource
-    m_pTrack->setType(m_pSoundSource->getType());
-
     if (metadataSynchronized) {
-        kLogger.info() << "Importing track metadata from file"
-                 << getUrl().toString();
+        // Metadata has been synchronized successfully at least
+        // once in the past. Only overwrite this information if
+        // new data has actually been imported, otherwise abort
+        // and preserve the existing data!
+        if (metadataImported.first == mixxx::MetadataSource::ImportResult::Succeeded) {
+            kLogger.info()
+                    << "Updating track metadata"
+                    << (pCoverImg ? "and embedded cover art" : "")
+                    << "from file"
+                    << getUrl().toString();
+        } else {
+            return; // abort
+        }
     } else {
-        kLogger.info() << "Initializing track metadata from file"
-                 << getUrl().toString();
+        DEBUG_ASSERT(pCoverImg);
+        kLogger.info()
+                << "Initializing track metadata and embedded cover art from file"
+                << getUrl().toString();
+    }
+
+    // Fallback: If artist or title fields are blank then try to populate
+    // them from the file name. This might happen if tags are unavailable,
+    // unreadable, or partially/completely missing.
+    // TODO(rryan): Should we re-visit this decision?
```

Comment:
I think we can remove this TODO. I am convinced that we are doing it right here.